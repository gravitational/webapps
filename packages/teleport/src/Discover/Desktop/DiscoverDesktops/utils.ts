import { Path } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/Path';
import { MonotoneX } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/MonotoneX';

const STROKE_WIDTH = 4;

export interface Line {
  width: number;
  height: number;
  path: string;
}

export function createLine(
  desktopServiceElement: HTMLDivElement,
  desktopElement: HTMLDivElement,
  containerElement: HTMLDivElement
): Line {
  if (!desktopElement || !desktopServiceElement || !containerElement) {
    return null;
  }

  const desktopServiceRect = desktopServiceElement.getBoundingClientRect();
  const desktopRect = desktopElement.getBoundingClientRect();
  const containerRect = containerElement.getBoundingClientRect();

  const distance = desktopRect.left - desktopServiceRect.right;

  const path = new Path();
  const line = new MonotoneX(path);

  line.lineStart();

  const desktopLinePosition =
    desktopRect.top - containerRect.top + desktopRect.height / 2 - 1;
  const desktopServiceLinePosition =
    desktopServiceRect.top - containerRect.top + desktopServiceRect.height / 2;

  line.point(0, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(40, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(distance - 10, desktopLinePosition + STROKE_WIDTH / 2);
  line.point(distance, desktopLinePosition + STROKE_WIDTH / 2);

  line.lineEnd();

  return {
    width: distance,
    height: containerRect.height,
    path: path.path,
  };
}
