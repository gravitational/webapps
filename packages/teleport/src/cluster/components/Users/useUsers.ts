/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { useState, useEffect } from 'react';
import { useAttempt } from 'shared/hooks';
import { User } from 'teleport/services/user';
import { useTeleport } from 'teleport/teleportContextProvider';

/**
 * useUsers contains state for Users view component.
 */
export default function useUsers() {
  const ctx = useTeleport();
  const [attempt, attemptActions] = useAttempt({ isProcessing: true });
  const [users, setUsers] = useState<User[]>([]);
  const [roles, setRoles] = useState<string[]>([]);
  const [action, setAction] = useState<Action>({
    type: 'none',
  });

  function onStartCreate() {
    setAction({ type: 'create' });
  }

  function onStartEdit(user: User) {
    setAction({ type: 'edit', user });
  }

  function onClose() {
    setAction({ type: 'none' });
  }

  function onSave(user: User, isNew = false) {
    return ctx.userService.upsertUser(user, isNew).then(response => {
      setUsers(users => [...users, response.user]);
      return response.token;
    });
  }

  function fetchRoles() {
    if (ctx.isRolesEnabled()) {
      return ctx.resourceService
        .fetchRoles()
        .then(resources => resources.map(role => role.name));
    }

    return Promise.resolve([]);
  }

  useEffect(() => {
    attemptActions.do(() =>
      Promise.all([fetchRoles(), ctx.userService.fetchUsers()]).then(values => {
        setRoles(values[0]);
        setUsers(values[1]);
      })
    );
  }, []);

  return {
    attempt,
    users,
    roles,
    action,
    onStartCreate,
    onStartEdit,
    onClose,
    onSave,
  };
}

type Action = {
  type: 'create' | 'edit' | 'none';
  user?: User;
};
