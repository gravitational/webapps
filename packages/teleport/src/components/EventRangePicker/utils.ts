import moment from 'moment';

export function getRangeOptions(): EventRange[] {
  return [
    {
      name: 'Today',
      from: moment(new Date())
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: '7 days',
      from: moment()
        .subtract(6, 'day')
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: 'Custom Range...',
      isCustom: true,
      from: new Date(),
      to: new Date(),
    },
  ];
}

export type EventRange = {
  from: Date;
  to: Date;
  isCustom?: boolean;
  name?: string;
};
