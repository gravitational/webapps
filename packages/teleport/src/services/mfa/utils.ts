/**
 * Copyright 2021 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Auth2faType, PreferredMfaType } from 'shared/services';

export type MfaOption = {
  value: Auth2faType;
  label: string;
};

export function getMfaOptions(
  mfa: Auth2faType,
  preferredMfa: PreferredMfaType,
  requireMfa = false
) {
  const mfaOptions: MfaOption[] = [];

  if (mfa === 'off' || !mfa) {
    return mfaOptions;
  }

  const mfaEnabled = mfa === 'on' || mfa === 'optional';
  const preferMfaWebauthn = mfaEnabled && preferredMfa === 'webauthn';
  const preferMfaU2f = mfaEnabled && preferredMfa === 'u2f';

  if (mfa === 'webauthn' || preferMfaWebauthn) {
    mfaOptions.push({ value: 'webauthn', label: 'Hardware Key' });
  }

  if (mfa === 'u2f' || preferMfaU2f) {
    mfaOptions.push({ value: 'u2f', label: 'Hardware Key' });
  }

  if (mfa === 'otp' || mfaEnabled) {
    mfaOptions.push({ value: 'otp', label: 'Authenticator App' });
  }

  if (!requireMfa && mfa === 'optional') {
    mfaOptions.push({ value: 'optional', label: 'None' });
  }

  return mfaOptions;
}
