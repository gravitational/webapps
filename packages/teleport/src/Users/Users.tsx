/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Indicator, Box, ButtonPrimary, Alert } from 'design';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import UserList from './UserList';
import UserAddEdit from './UserAddEdit';
import UserDelete from './UserDelete';
import UserReset from './UserReset';
import useUsers, { State } from './useUsers';

export default function Container() {
  const state = useUsers();
  return <Users {...state} />;
}

export function Users(props: State) {
  const {
    attempt,
    users,
    roles,
    operation,
    onStartCreate,
    onStartDelete,
    onStartEdit,
    onStartReset,
    onClose,
    onCreate,
    onUpdate,
    onDelete,
    onReset,
  } = props;

  return (
    <FeatureBox>
      <FeatureHeader>
        <FeatureHeaderTitle>Users</FeatureHeaderTitle>
        {attempt.isSuccess && (
          <ButtonPrimary ml="auto" width="240px" onClick={onStartCreate}>
            Create New User
          </ButtonPrimary>
        )}
      </FeatureHeader>
      {attempt.isProcessing && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {attempt.isFailed && <Alert kind="danger" children={attempt.message} />}
      {attempt.isSuccess && (
        <UserList
          users={users}
          onEdit={onStartEdit}
          onDelete={onStartDelete}
          onReset={onStartReset}
        />
      )}
      {(operation.type === 'create' || operation.type === 'edit') && (
        <UserAddEdit
          isNew={operation.type === 'create'}
          roles={roles}
          onClose={onClose}
          onCreate={onCreate}
          onUpdate={onUpdate}
          user={operation.user}
        />
      )}
      {operation.type === 'delete' && (
        <UserDelete
          onClose={onClose}
          onDelete={onDelete}
          username={operation.user.name}
        />
      )}
      {operation.type === 'reset' && (
        <UserReset
          onClose={onClose}
          onReset={onReset}
          username={operation.user.name}
        />
      )}
    </FeatureBox>
  );
}
