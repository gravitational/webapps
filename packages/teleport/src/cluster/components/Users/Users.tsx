/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import { Indicator, Box, ButtonPrimary, Alert } from 'design';
import UserList from './UserList';
import UserDialog from './UserDialog/UserDialog';
import useUsers from './useUsers';
import { useTeleport } from 'teleport/teleportContextProvider';

export default function Container() {
  const state = useUsers();

  return <Users state={state} />;
}

/**
 * Users is a view component that:
 * - Displays list of users stored in a backend.
 * - Provide user related actions: edit/view, delete, reset
 *   based on permissions.
 */
export function Users({ state }: Props) {
  const {
    attempt,
    users,
    roles,
    operation,
    onStartCreate,
    onStartEdit,
    onClose,
    onSave,
  } = state;

  const ctx = useTeleport().storeUser;
  const canCreate =
    attempt.isSuccess && ctx.getUserAccess().create && ctx.getRoleAccess().read;

  return (
    <FeatureBox>
      <FeatureHeader>
        <FeatureHeaderTitle>Users</FeatureHeaderTitle>
        {canCreate && (
          <ButtonPrimary ml="auto" width="240px" onClick={onStartCreate}>
            Add User
          </ButtonPrimary>
        )}
      </FeatureHeader>
      {attempt.isProcessing && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {attempt.isFailed && <Alert kind="danger" children={attempt.message} />}
      {attempt.isSuccess && (
        <UserList users={users} pageSize={20} onEdit={onStartEdit} />
      )}
      {(operation.type === 'create' || operation.type === 'edit') && (
        <UserDialog
          roles={roles}
          onClose={onClose}
          onSave={onSave}
          user={operation.user}
        />
      )}
    </FeatureBox>
  );
}

type Props = {
  state: ReturnType<typeof useUsers>;
};
