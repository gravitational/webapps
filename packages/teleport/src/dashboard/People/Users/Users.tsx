/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import { Indicator, Box, ButtonPrimary, Alert } from 'design';
import UserList from './UserList';
import UserAddEdit from './UserAddEdit';
import UserDelete from './UserDelete';
import UserReset from './UserReset';
import useUsers from './useUsers';
import { useTeleport } from 'teleport/teleportContextProvider';

export default function Container() {
  const state = useUsers();
  return <Users {...state} />;
}

/**
 * Users is a view component that:
 * - Displays list of users stored in a backend.
 * - Provide user related actions: edit/view, delete, reset
 *   based on permissions.
 */
export function Users(props: ReturnType<typeof useUsers>) {
  const {
    attempt,
    users,
    roles,
    operation,
    onStartCreate,
    onStartDelete,
    onStartEdit,
    onStartReset,
    onClose,
    onCreate,
    onUpdate,
    onDelete,
    onReset,
  } = props;

  const ctx = useTeleport().storeUser;
  const canCreate = attempt.isSuccess && ctx.getUserAccess().create;
  const canDelete = ctx.getUserAccess().remove;
  const canUpdate = ctx.getUserAccess().edit;

  return (
    <FeatureBox>
      <FeatureHeader>
        <FeatureHeaderTitle>Users</FeatureHeaderTitle>
        {canCreate && (
          <ButtonPrimary ml="auto" width="240px" onClick={onStartCreate}>
            Create New User
          </ButtonPrimary>
        )}
      </FeatureHeader>
      {attempt.isProcessing && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {attempt.isFailed && <Alert kind="danger" children={attempt.message} />}
      {attempt.isSuccess && (
        <UserList
          users={users}
          pageSize={20}
          onEdit={onStartEdit}
          onDelete={onStartDelete}
          onReset={onStartReset}
          canDelete={canDelete}
          canUpdate={canUpdate}
        />
      )}
      {(operation.type === 'create' || operation.type === 'edit') && (
        <UserAddEdit
          isNew={operation.type === 'create'}
          roles={roles}
          onClose={onClose}
          onCreate={onCreate}
          onUpdate={onUpdate}
          user={operation.user}
        />
      )}
      {operation.type === 'delete' && (
        <UserDelete
          onClose={onClose}
          onDelete={onDelete}
          username={operation.user.name}
        />
      )}
      {operation.type === 'reset' && (
        <UserReset
          onClose={onClose}
          onReset={onReset}
          username={operation.user.name}
        />
      )}
    </FeatureBox>
  );
}
