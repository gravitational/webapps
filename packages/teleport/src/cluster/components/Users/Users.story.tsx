/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import Users from './Users';
import resourceService from 'e-teleport/services/resources';
import userServices from 'teleport/services/user';
import StoryContextProvider, {
  dummyCtx,
} from 'design/utils/StoryTeleportContextProvider';

export default {
  title: 'Teleport/Users/UsersView',
};

export const Success = () => {
  resourceService.fetchRoles = () =>
    Promise.resolve([
      {
        content: '',
        displayName: '',
        id: '',
        kind: 'role',
        name: 'admin',
      },
      {
        content: '',
        displayName: '',
        id: '',
        kind: 'role',
        name: 'testrole',
      },
    ]);
  userServices.fetchUsers = () => Promise.resolve([]);
  return (
    <StoryContextProvider ctx={dummyCtx}>
      <Users />
    </StoryContextProvider>
  );
};

export const Processing = () => {
  resourceService.fetchRoles = () => new Promise(() => null);
  userServices.fetchUsers = () => new Promise(() => null);
  return (
    <StoryContextProvider ctx={dummyCtx}>
      <Users />
    </StoryContextProvider>
  );
};

export const Failed = () => {
  resourceService.fetchRoles = () =>
    Promise.reject(new Error('some error message'));
  return (
    <StoryContextProvider ctx={dummyCtx}>
      <Users />
    </StoryContextProvider>
  );
};

userServices.createUser = () => new Promise(() => null);
