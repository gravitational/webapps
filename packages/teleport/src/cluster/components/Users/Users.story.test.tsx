import React from 'react';
import { Success, Processing, Failed } from './Users.story';
import { render } from 'design/utils/testing';
import { Logger } from 'shared/libs/logger';

test('success state', async () => {
  const { container, findByText } = render(<Success />);
  await findByText(/no data available/i);
  expect(container).toMatchSnapshot();
});

test('processing state', async () => {
  const { container, findByTestId } = render(<Processing />);
  await findByTestId('loader');
  expect(container.firstChild).toMatchSnapshot();
});

// TODO: uncomment after we remove throw from useAttempt run().
// Currently, this test will fail from unhandled promise rejections.
// test('failed state', async () => {
//   jest.spyOn(Logger.prototype, 'error').mockImplementation();
//   const { container, findByText } = render(<Failed />);

//   await findByText(/error/i);

//   expect(container).toMatchSnapshot();
// });
