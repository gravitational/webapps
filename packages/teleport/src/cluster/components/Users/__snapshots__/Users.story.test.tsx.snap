// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processing state 1`] = `
.c3 {
  box-sizing: border-box;
  margin: 72px;
  text-align: center;
}

.c4 {
  display: inline-block;
  transition: color .3s;
  color: #FFFFFF;
}

.c5 {
  font-size: 32px;
  height: 32px;
  width: 32px;
  animation: anim-rotate 1s infinite linear;
  color: #fff;
  display: inline-block;
  margin: 16px;
  opacity: 0.24;
}

.c1 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  flex-shrink: 0;
  border-bottom: 1px solid #1C254D;
  height: 56px;
  margin-left: -40px;
  margin-right: -40px;
  padding-left: 40px;
  padding-right: 40px;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  margin: 0px;
  flex-shrink: 0;
}

.c0 {
  box-sizing: border-box;
  padding-left: 40px;
  padding-right: 40px;
  display: flex;
  overflow: auto;
  width: 100%;
  height: 100%;
  flex-direction: column;
}

.c0::after {
  content: ' ';
  padding-bottom: 24px;
}

<div
  class="sc-fzqBZW c0"
>
  <div
    class="sc-fzqBZW c1"
  >
    <div
      class="sc-fznZeY c2"
    >
      Users
    </div>
  </div>
  <div
    class="c3"
  >
    <span
      class="c4 icon icon-spinner8 c5"
      color="light"
      data-testid="loader"
    />
  </div>
</div>
`;

exports[`success state 1`] = `
.c3 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  margin-left: auto;
  width: 240px;
}

.c3:active {
  opacity: 0.56;
}

.c3:hover,
.c3:focus {
  background: #651FFF;
}

.c3:active {
  background: #354AA4;
}

.c3:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c10 {
  display: inline-block;
  transition: color .3s;
  color: #FFFFFF;
  font-size: 16px;
}

.c12 {
  display: inline-block;
  transition: color .3s;
  color: #FFFFFF;
}

.c14 {
  display: inline-block;
  transition: color .3s;
  margin-right: 8px;
  color: #FFFFFF;
}

.c7 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 400;
  font-size: 12px;
  line-height: 16px;
  margin: 0px;
  color: #FFFFFF;
}

.c13 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 16px;
  line-height: 32px;
  margin: 24px;
  color: rgba(255,255,255,0.87);
}

.c4 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
}

.c1 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  flex-shrink: 0;
  border-bottom: 1px solid #1C254D;
  height: 56px;
  margin-left: -40px;
  margin-right: -40px;
  padding-left: 40px;
  padding-right: 40px;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  margin: 0px;
  flex-shrink: 0;
}

.c0 {
  box-sizing: border-box;
  padding-left: 40px;
  padding-right: 40px;
  display: flex;
  overflow: auto;
  width: 100%;
  height: 100%;
  flex-direction: column;
}

.c0::after {
  content: ' ';
  padding-bottom: 24px;
}

.c11 {
  background: #222C59;
  box-shadow: 0 4px 16px rgba(0,0,0,0.24);
  border-collapse: collapse;
  border-spacing: 0;
  border-radius: 8px;
  font-size: 12px;
  width: 100%;
}

.c11 > thead > tr > th,
.c11 > tbody > tr > th,
.c11 > tfoot > tr > th,
.c11 > thead > tr > td,
.c11 > tbody > tr > td,
.c11 > tfoot > tr > td {
  padding: 8px 8px;
  vertical-align: middle;
}

.c11 > thead > tr > th:first-child,
.c11 > tbody > tr > th:first-child,
.c11 > tfoot > tr > th:first-child,
.c11 > thead > tr > td:first-child,
.c11 > tbody > tr > td:first-child,
.c11 > tfoot > tr > td:first-child {
  padding-left: 24px;
}

.c11 > thead > tr > th:last-child,
.c11 > tbody > tr > th:last-child,
.c11 > tfoot > tr > th:last-child,
.c11 > thead > tr > td:last-child,
.c11 > tbody > tr > td:last-child,
.c11 > tfoot > tr > td:last-child {
  padding-right: 24px;
}

.c11 > thead > tr > th {
  background: #111B48;
  color: #FFFFFF;
  cursor: pointer;
  font-size: 10px;
  font-weight: 400;
  padding-bottom: 0;
  padding-top: 0;
  text-align: left;
  opacity: 0.75;
  text-transform: uppercase;
  white-space: nowrap;
}

.c11 > thead > tr > th .c9 {
  font-weight: bold;
  font-size: 8px;
  margin-left: 8px;
}

.c11 > tbody > tr > td {
  color: rgba(255,255,255,0.87);
  line-height: 16px;
}

.c11 tbody tr {
  border-bottom: 1px solid #1C254D;
}

.c11 tbody tr:hover {
  background-color: rgb(37,49,98);
}

.c8 {
  box-sizing: border-box;
  display: flex;
}

.c8 button {
  background: none;
  border: none;
  border-radius: 200px;
  cursor: pointer;
  padding: 0;
  margin: 0 0 0 8px;
  outline: none;
  transition: all 0.3s;
  text-align: center;
}

.c8 button:hover,
.c8 button:focus {
  background: #1C254D;
}

.c8 button:hover .c9,
.c8 button:focus .c9 {
  opacity: 1;
}

.c8 button .c9 {
  opacity: 0.4;
  font-size: 20px;
  transition: all 0.3s;
}

.c6 {
  padding: 8px 24px;
  display: flex;
  height: 24px;
  flex-shrink: 0;
  align-items: center;
  justify-content: space-between;
  background: #222C59;
  border-top-left-radius: 8px;
  border-top-right-radius: 8px;
}

.c5 {
  box-sizing: border-box;
  font-size: 12px;
  min-width: 200px;
  outline: none;
  border: none;
  border-radius: 200px;
  height: 32px;
  transition: all .2s;
  background: #222C59;
  margin-right: 16px;
  padding-left: 16px;
  padding-right: 16px;
  color: rgba(255,255,255,0.87);
  height: 30px;
}

.c5:hover {
  background: #2C3A73;
}

.c5:focus,
.c5:active {
  background: #2C3A73;
  box-shadow: inset 0 2px 4px rgba(0,0,0,.24);
  color: rgba(255,255,255,0.87);
}

.c5::placeholder {
  color: rgba(255,255,255,0.24);
  font-size: 12px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Users
      </div>
      <button
        class="c3"
        kind="primary"
        width="240px"
      >
        Add User
      </button>
    </div>
    <div>
      <div
        class="c4"
      >
        <input
          class="c5"
          color="text.primary"
          height="30px"
          placeholder="SEARCH..."
          value=""
        />
      </div>
      <div>
        <nav
          class="c6"
        >
          <div
            class="c7"
            color="primary.contrastText"
          >
            SHOWING 
            <strong>
              0
            </strong>
             - 
            <strong>
              0
            </strong>
             
            of 
            <strong>
              0
            </strong>
          </div>
          <div
            class="c8"
          >
            <button
              disabled=""
              title="Previous Page"
            >
              <span
                class="c9 c10 icon icon-arrow-left-circle "
                color="light"
                font-size="3"
              />
            </button>
            <button
              disabled=""
              title="Next Page"
            >
              <span
                class="c9 c10 icon icon-arrow-right-circle "
                color="light"
                font-size="3"
              />
            </button>
          </div>
        </nav>
        <table
          class="c11"
        >
          <thead>
            <tr>
              <th>
                <a>
                  Username
                  <span
                    class="c9 c12 icon icon-chevrons-expand-vertical "
                    color="light"
                  />
                </a>
              </th>
              <th>
                <a>
                  Roles
                  <span
                    class="c9 c12 icon icon-chevrons-expand-vertical "
                    color="light"
                  />
                </a>
              </th>
              <th>
                <a>
                  Created
                  <span
                    class="c9 c12 icon icon-chevron-up "
                    color="light"
                  />
                </a>
              </th>
              <th />
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                colspan="4"
              >
                <div
                  class="c13"
                  color="text.primary"
                  style="display: flex; align-items: center; justify-content: center;"
                >
                  <span
                    class="c9 c14 icon icon-database "
                    color="light"
                  />
                  No Data Available
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
